alphazero/model.py

```diff
@@ -0,0 +1,177 @@
+            in_channels: # of game states(previous turns and positions)
+            enc_channels: number of channels to encode
+            num_blocks: number of residual blocks to create
+    """
+
+    def __init__(
+        self,
+        block,
+        in_channels: int,
+        enc_channels: int,
+        num_blocks: int,
+        activation=nn.ReLU(inplace=False),
+        batch_on=True,
+    ):
+        super(SidNet, self).__init__()
+        self.conv1 = nn.Conv2d(
+            in_channels, enc_channels, kernel_size=3, stride=1, padding=1
+        )
+        self.bn = nn.BatchNorm2d(enc_channels) if batch_on else None
+        self.resblock = self.make_layer(block, enc_channels, num_blocks)
+        self.activation = activation
+
+    def make_layer(self, block, out_channels, num_blocks):
+
+        layers = []
+        for i in range(0, num_blocks):
+            layers.append(block(out_channels, out_channels))
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
+
+        encoding = self.conv1(x)
+        if self.bn is not None:
+            encoding = self.bn(encoding)
+        encoding = self.activation(encoding)
+        encoding = self.resblock(encoding)
+
+        return encoding
+
+
+class PolicyHead(nn.Module):
+    """
+    Create the PolicyHead Network, which take the encoded features and the game
+    being played from the previous SidNet Network and returns a vector of
+    length game.width * game.height of move probabilities for each specific
+    position. Following the architecture of the AlphaGoZero Paper, requires the
+    number of channels from the previous residual nets, the game being played,
+    the activation function, and if bathcnorm is on or off.
+    """
+
+    def __init__(
+        self,
+        enc_channels: int,
+        game,
+        activation=nn.ReLU(inplace=False),
+        batch_on=True,
+        pass_move=True,
+    ):
+
+        super(PolicyHead, self).__init__()
+        self.policyconv = nn.Conv2d(
+            enc_channels, out_channels=2, kernel_size=1, stride=1
+        )
+        self.policybn = nn.BatchNorm2d(2) if batch_on else None
+        self.poss_moves = game.width * game.height
+
+        if pass_move:
+            self.policyfc = nn.Linear(
+                in_features=self.poss_moves * 2,
+                out_features=self.poss_moves + 1,
+            )
+        else:
+            self.policyfc = nn.Linear(
+                in_features=self.poss_moves * 2, out_features=self.poss_moves
+            )
+        self.activation = activation
+        self.logsoftmax = nn.LogSoftmax(dim=1)
+
+    def forward(self, encoding):
+        policy = self.policyconv(encoding)
+        if self.policybn is not None:
+            policy = self.policybn(policy)
+        policy = self.activation(policy)
+        policy = policy.view(-1, self.poss_moves * 2)
+        policy = self.policyfc(policy)
+        probas = self.logsoftmax(policy).exp()
```

Comment:
It it possible to reshape back before returning here? For easier interaction with game api later

Reply:
Will do that!